From 7fdb9cad2c3e5a82902f814037899d3cc7dcc9f8 Mon Sep 17 00:00:00 2001
From: Igor Grinberg <igrinberg@habana.ai>
Date: Tue, 07 Apr 2020 09:17:38 +0300
Subject: [PATCH] [SW-11180] preboot: boot_desc: update pending reg

To be backward compatible, we need to update the "update pending
register" also for the new boot way.
(re)use the preboot_set_upd_pending_reg() function for this task.

Change-Id: Ie37383945c6bf3cfde43153e654b24b96b63ee30
Signed-off-by: Igor Grinberg <igrinberg@habana.ai>
(cherry picked from commit f5acb4d119fe24ac7c9e2fca191ad913e789727f)
---

diff --git a/app/ppboot.c b/app/ppboot.c
index 3b34975..1dbcddf 100644
--- a/app/ppboot.c
+++ b/app/ppboot.c
@@ -50,10 +50,11 @@
 {
 	u32 upd_pending_reg = get_upd_pending_sts_reg();
 	u32 image_pending = regmap_readl(upd_pending_reg);
-	u32 off;
+	u32 off, idx;
 
 	if (boot_desc_init()) {
-		off = boot_desc_get_offset(image_pending, HL_FW_IMG_PREBOOT);
+		off = boot_desc_get_offset(image_pending,
+					   HL_FW_IMG_PREBOOT, &idx);
 		if (off != HL_FW_IMG_NOT_FOUND) {
 			*offset = off;
 			return image_pending;
diff --git a/app/preboot.c b/app/preboot.c
index 97490e2..3d4bf1f 100644
--- a/app/preboot.c
+++ b/app/preboot.c
@@ -90,12 +90,14 @@
 {
 	u32 upd_pending_reg = get_upd_pending_sts_reg();
 	u32 image_pending = regmap_readl(upd_pending_reg);
-	u32 offset;
+	u32 offset, idx;
 
 	if (boot_desc_init()) {
-		offset = boot_desc_get_offset(image_pending, HL_FW_IMG_UBOOT);
+		offset = boot_desc_get_offset(image_pending,
+					      HL_FW_IMG_UBOOT, &idx);
 		if (offset != HL_FW_IMG_NOT_FOUND) {
 			*pending = image_pending;
+			preboot_set_upd_pending_reg(idx);
 			return offset;
 		}
 	}
diff --git a/common/boot_desc.c b/common/boot_desc.c
index ed3258c..0b080b5 100644
--- a/common/boot_desc.c
+++ b/common/boot_desc.c
@@ -86,7 +86,7 @@
 }
 
 u32 boot_desc_get_offset(enum fw_post_reset_cmd upd_status,
-			 enum hl_fw_img_type type)
+			 enum hl_fw_img_type type, u32 *ridx)
 {
 	int i, idx;
 
@@ -100,7 +100,7 @@
 			idx = (idx + 1) % 2;
 
 		preboot_puts_hex(" - found! using index: ", idx, false);
-
+		*ridx = idx;
 		return img_desc[i].offset[idx];
 	}
 
diff --git a/common/boot_desc.h b/common/boot_desc.h
index b64e0f1..e323e05 100644
--- a/common/boot_desc.h
+++ b/common/boot_desc.h
@@ -26,12 +26,13 @@
  * @upd_status	- FLOPS_UPD_PENDING - for modifiable image offset,
  *		  other values - for stable image offset
  * @type	- the image type to look for
+ * @idx		- pointer for the new index
  *
  * Returns:
  * On success: the offset to the requested image
  * On failure: (u32)-1 (no image by the requested type)
  */
 u32 boot_desc_get_offset(enum fw_post_reset_cmd upd_status,
-			 enum hl_fw_img_type type);
+			 enum hl_fw_img_type type, u32 *idx);
 
 #endif /* _BOOT_DESC_H_ */
